daemon: Track policy implementation delay by source

raybejjani opened the pull request:
We see high variance in the time from a .PolicyAdd in the daemon to when
a revision is finally in the datapath. Quantifying this delay is
valuable when diagnosing certain classes of problems, particularly
transient drops to IPs seen by the DNS Proxy and inserted into node's
policy.
The options to .PolicyAdd now includes a source field that reflects the
source of that specific invocation. It is expected to be one of api,
fqdn or k8s. This information is passed onto a new goroutine that waits
on each revision to be deployed and records the time since the
.PolicyAdd invocation into a new policy_implementation_delay histogram.

I ended up going with a callback from each endpoint. The code was a little smaller and the timings more precise. The global max represents the worst case. The downside is that regenerations from different revisions are not distinguishable.

The data is a histogram, and can be plotted over time with a heatmap:
![image](https://user-images.githubusercontent.com/744312/53513318-28dc3800-3ac5-11e9-863b-5e36d83e4af6.png)



The JSON for the panel is
<details>
{
  "type": "heatmap",
  "title": "PolicyAdd Implementation Delay",
  "gridPos": {
    "x": 0,
    "y": 0,
    "w": 12,
    "h": 9
  },
  "id": 119,
  "datasource": "prometheus",
  "targets": [
    {
      "refId": "A",
      "expr": "sum(increase(cilium_policy_implementation_delay_bucket[1m])) by (le)",
      "intervalFactor": 1,
      "format": "heatmap",
      "legendFormat": "{{le}}"
    }
  ],
  "heatmap": {},
  "cards": {
    "cardPadding": null,
    "cardRound": null
  },
  "color": {
    "mode": "opacity",
    "cardColor": "#b4ff00",
    "colorScale": "sqrt",
    "exponent": 0.5,
    "colorScheme": "interpolateOranges"
  },
  "legend": {
    "show": false
  },
  "dataFormat": "tsbuckets",
  "yBucketBound": "auto",
  "xAxis": {
    "show": true
  },
  "yAxis": {
    "show": true,
    "format": "s",
    "decimals": null,
    "logBase": 1,
    "splitFactor": null,
    "min": null,
    "max": null
  },
  "xBucketSize": null,
  "xBucketNumber": null,
  "yBucketSize": null,
  "yBucketNumber": null,
  "tooltip": {
    "show": true,
    "showHistogram": true
  },
  "highlightCards": true,
  "links": []
}</details>



<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7160)
<!-- Reviewable:end -->

raybejjani committed:
daemon: Track policy implementation delay by source

We see high variance in the time from a .PolicyAdd in the daemon to when
a revision is finally in the datapath. Quantifying this delay is
valuable when diagnosing certain classes of problems, particularly
transient drops to IPs seen by the DNS Proxy and inserted into node's
policy.
The options to .PolicyAdd now includes a source field that reflects the
source of that specific invocation. It is expected to be one of api,
fqdn or k8s. This information is passed onto a callback that will be
called when each endpoint reaches the revision in the .PolicyAdd that
created the revision. This information is recorded into a new
policy_implementation_delay histogram metric for each endpoint.

Signed-off-by: Ray Bejjani <ray@covalent.io>

tgraf commented on daemon/metrics.go:
```diff
@@ -134,3 +137,117 @@ func init() {
+			// (WaitForEndpointsAtPolicyRevision returns when there are no
+			// endpoints).
+			if len(endpointmanager.GetEndpoints()) == 0 {
+				time.Sleep(10 * time.Second)
```
This seems long. Just wait one second so we have metrics soon one the first endpoint is added. I'd rather add a function `endpointmanager.NumEndpoints()` that we can optimize than optimizing on the interval.

raybejjani replied:
👍 (but the callback implementation doesn't have this goroutine at all, and avoids this problem)

tgraf commented on daemon/metrics.go:
```diff
@@ -134,3 +137,117 @@ func init() {
+					case !stillOpen:
+						return
+					case start.rev <= currentRevision:
+						duration := now.Sub(start.timestamp)
```
Does this need to use `spanstat` to not run into the negative metric problem?

raybejjani replied:
Good point. I added a guard but I think histograms won't panic (it was a counter the last time and those are supposed to only go up).

tgraf reviewed (changes requested):
Looks great overall

raybejjani committed:
endpoints: Add optional callback to WaitForPolicyRevision

Endpoints already have a notification scheme for when they reach certain
policy revisions. This is convenient for setting true/false values like
the Enforcing field on CNPs. It is more difficult to measure when the
revision is reached by the endpoint because the goroutine waiting on the
channel returned by WaitForPolicyRevision may not be scheduled
immediately. Using a callback means arbitrary code can be run, ideally
non-blocking. We also pass in the timestamp to ensure a more accurate
time measurement is possible.

Signed-off-by: Ray Bejjani <ray@covalent.io>

vadorovsky commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -343,7 +343,7 @@ func (s *EndpointSuite) TestWaitForPolicyRevision(c *C) {
 	e.state = StateCreating
 	ctx, cancel = context.WithCancel(context.Background())
-	ch = e.WaitForPolicyRevision(ctx, 99)
+	ch = e.WaitForPolicyRevision(ctx, 99, nil)
```
Maybe we should add a test case for `WaitForPolicyRevision` with some non-nil `done` function?

raybejjani replied:
@mrostecki Good point :) I shouldn't have been lazy. I added a check in the same test, since the callback should be called whenever the channel is closed. Let me know what you think, I left comments but they aren't really that useful :/

vadorovsky reviewed (commented):
Looks good! I have only one suggestion for a unit test.

tgraf reviewed (approved):
Awesome
